src/ml.py

```diff
@@ -70,10 +97,58 @@ def encode_batch(
         Returns:
             np.ndarray
         """
+        if experimental_use_sliding_window:
+            return self._encode_batch_using_sliding_window(
+                text_batch, batch_size=batch_size, device=device
+            )
+
         return self.encoder.encode(
             text_batch, batch_size=batch_size, show_progress_bar=False, device=device
         )
 
+    def _encode_batch_using_sliding_window(
+        self, text_batch: list[str], batch_size: int = 32, device: Optional[str] = None
+    ):
+        """
+        Encode a batch of strings accommodating long texts using a sliding window.
+
+        For args, see encode_batch.
+        """
+
+        max_seq_length = self.encoder.max_seq_length
+        assert isinstance(max_seq_length, int)
+
+        # Split the texts based on length and apply sliding window only to longer texts
+        processed_texts = []
+        window_lengths = []
+
+        for text in text_batch:
+            if self.get_n_tokens(text) > max_seq_length:
+                windows = sliding_window(
+                    text, window_size=max_seq_length, stride=max_seq_length // 2
+                )  # Use max_seq_length as window size and half of it as stride
+                processed_texts.extend(windows)
+                window_lengths.append(len(windows))
+            else:
+                processed_texts.append(text)
+                window_lengths.append(1)
+
+        embeddings = self.encode_batch(
+            processed_texts, batch_size=batch_size, device=device
+        )
+
+        reduced_embeddings = []
+
+        for length in window_lengths:
+            if length > 1:
+                reduced_embeddings.append(np.mean(embeddings[:length], axis=0))
+                embeddings = embeddings[length:]
+            else:
+                reduced_embeddings.append(embeddings[0])
+                embeddings = embeddings[1:]
+
+        return np.vstack(reduced_embeddings)
+
```

Comment:
I like this -- makes more sense now on the sliding-window question, we're splitting text that is too long, embedding and then aggregating by averaging. 

One thing we should be cautious of here is the averaging: it might occur, that `self.get_n_tokens(text) - max_seq_length` is small, let's say a few tokens, in which case we're averaging an embedding of most of the text, and that of a single word.

A thought: if windows are already larger than max context length, would it be bad to recreate windows at this point instead of glueing them back together? I.e. the sliding window function spits out the new windows, and we use them going forward (running the risk of them not beeing split nicely at sentence-ends). This might bring some regularisation to passage sizes at the price of abandoning the text-block borders.

Reply:
I don't think we're averaging the embedding of most of the text and a single word, because the window size is always `max_seq_length`? I know it's not ideal, but I figured this is the solution that is most used in the literature and presumably better than ignoring text past the context window length.

I like the windows -> new text blocks idea, but I think this falls into a broader "what's a better way to model text" conversation that would require Vespa schema upgrades and product changes. I'm pretty sure context window length of the retrieval model will come up in those chats when we have them